### PR TITLE
fix `prepare_inputs_for_generation` logic for Prompt Learning methods

### PR DESCRIPTION
**Pull request by pacman100**
### What does this PR do?
1. The current logic from PR https://github.com/huggingface/peft/pull/1252 is incorrect as explained in https://github.com/huggingface/peft/pull/1252#issuecomment-1888604086 and https://github.com/huggingface/peft/pull/1252#issuecomment-1888614334.
2. This PR fixes it.

### CODE REVIEW

**Comment by gante** [comment]
(LGTM)

**Review by BenjaminBossan — APPROVED** [review]
Thanks for doing the hard work and figuring this out. LGTM.

Do you have an example where the current logic fails. We could add it as a test (in a later PR) to prevent regressions.

**Comment by pacman100** [comment]
> Do you have an example where the current logic fails. 

The current logic doesn't fail in the sense of runtime errors but will produce wrong outputs as the initial input ids are again run through the model and they involve in self-attention with themselves. Will look into test in a follow up PR.

**Comment by BenjaminBossan** [comment]
> The current logic doesn't fail in the sense of runtime errors but will produce wrong outputs as the initial input ids are again run through the model and they involve in self-attention with themselves. Will look into test in a follow up PR.

Ah, got it, thanks. That makes it indeed very tricky to test.

**Comment by gante** [comment]
> The current logic doesn't fail in the sense of runtime errors but will produce wrong outputs as the initial input ids are again run through the model and they involve in self-attention with themselves. 

@pacman100 @BenjaminBossan for these cases, in generate-related tasks in `transformers`, we often test the some of the returned logits against hardcoded expected values (e.g. [here](https://github.com/huggingface/transformers/blob/07bdbebb48a9fe1e748348e4e14ae0b4659e54c4/tests/models/mistral/test_modeling_mistral.py#L480)). In some cases, when the point is to test the logic (and not so much a model architecture), we use a tiny dummy model.
